packages/core/src/shared/utilities/textDocumentUtilities.ts

```diff
@@ -76,3 +79,97 @@ export function getTabSize(editor?: vscode.TextEditor): number {
+        new vscode.Position(selection.end.line, doc.lineAt(selection.end.line).range.end.character)
+    )
+}
+
+/**
+ * Applies the given code to the specified range in the document.
+ * Saves the document after the edit is successfully applied.
+ *
+ * @param {vscode.TextDocument} doc - The VSCode document to which the changes are applied.
+ * @param {vscode.Range} range - The range in the document where the code is replaced.
+ * @param {string} code - The code to be applied to the document.
+ * @returns {Promise<void>} - Resolves when the changes are successfully applied and the document is saved.
+ */
+export async function applyChanges(doc: vscode.TextDocument, range: vscode.Range, code: string) {
+    const edit = new vscode.WorkspaceEdit()
+    edit.replace(doc.uri, range, code)
+    const successfulEdit = await vscode.workspace.applyEdit(edit)
+    if (successfulEdit) {
+        getLogger().debug('Diff: Edits successfully applied to: %s', doc.uri.fsPath)
+        await doc.save()
+    } else {
+        getLogger().error('Diff: Unable to apply changes to: %s', doc.uri.fsPath)
+    }
+}
+
+/**
+ * Creates a temporary file for diff comparison by cloning the original file
+ * and applying the proposed changes within the selected range.
+ *
+ * @param {vscode.Uri} originalFileUri - The URI of the original file.
+ * @param {any} message - The message object containing the proposed code changes.
+ * @param {vscode.Selection} selection - The selection range in the document where the changes are applied.
+ * @returns {Promise<vscode.Uri>} - A promise that resolves to the URI of the temporary file.
+ */
+export async function createTempFileForDiff(
+    originalFileUri: vscode.Uri,
+    message: any,
+    selection: vscode.Selection,
+    scheme: string
+): Promise<vscode.Uri> {
+    const errorCode = 'createTempFile'
+    const id = Date.now()
+    const languageId = (await vscode.workspace.openTextDocument(originalFileUri)).languageId
+    const tempFile = _path.parse(originalFileUri.path)
+    const tempFilePath = _path.join(tempDirPath, `${tempFile.name}_proposed-${id}${tempFile.ext}`)
+    try {
+        if (!(await fs.exists(tempDirPath))) {
+            await fs.mkdir(tempDirPath)
+        }
+    } catch (error) {
+        getLogger().error('Diff: Unable to create directory %s: %s', tempDirPath, (error as Error).message)
+    }
+
+    const tempFileUri = vscode.Uri.parse(`${scheme}:${tempFilePath}`)
+    getLogger().debug('Diff: Creating temp file: %s', tempFileUri.fsPath)
+
+    try {
+        // Write original content to temp file
+        await fs.writeFile(tempFilePath, await fs.readFileAsString(originalFileUri.fsPath))
+    } catch (error) {
+        if (!(error instanceof Error)) {
+            throw error
+        }
+        getLogger().error('Diff: Unable to write to file %s: %s', tempFilePath, (error as Error).message)
+        ToolkitError.chain(error, 'Unable to write to file', { code: errorCode })
```

Comment:
add 'throw'. Also we want this error message to give context to why the FS call failed. So in this case it should be "Failed to write to temp file"